torch/csrc/jit/passes/guard_elimination.cpp

```diff
@@ -122,36 +148,85 @@ struct GuardElimination {
     }
   }
 
- private:
-  bool removableGuard(Node* n) {
-    if (!simple_ops_.count(n->kind())) {
-      return false;
-    }
-
+  bool checkInputs(Node *n, const std::unordered_set<size_t> &except) {
     bool all_inputs_guarded = true;
+    size_t i = 0;
     for (auto input : n->inputs()) {
       if (input->node()->kind() == prim::Guard ||
-          input->node()->kind() == prim::Constant) {
+          input->node()->kind() == prim::Constant ||
+          input->type()->isSubtypeOf(NumberType::get()) ||
+          except.count(i) != 0) {
         AT_ASSERT(
             input->node()->kind() != prim::Guard ||
             input->type()->expect<TensorType>());
       } else {
+        GRAPH_DEBUG("input ", input->debugName(), " isn't guarded, type ",
+                    *input->type());
         all_inputs_guarded = false;
         break;
       }
+      i++;
     }
     return all_inputs_guarded;
   }
 
+private:
+  // `removableGuard` checks if it is valid to remove the `prim::Guard`
+  // guarding the output of some operation.
+  // For a large number of operations, the test is to simply check
+  // that all inputs to the operation are also guarded.
+  bool removableGuard(Node *n) {
```

Comment:
There are some implicit assumptions in this guard removal pass that are not document here. In particular this pass does not work generically, it only works on the results of a profiling run where we conservatively merge TensorTypes across runs. An  example:
```
a = guard(FloatTensor, _a)
b = guard(FloatTensor, _b)
c = guard(Float(3, 4), a + b) # this guard is not redundant!
```
However, we know that profiling will not generate this case, ie. the guards are consistent with profiling. This has to be documented somewhere or it will get broken. Furthermore, _every pass_ that exists between when the profiling is generated and when this pass runs must maintain pretty fragile invariants that this pass expects. Is there anyway to run redundant guard removal earlier? It seems to be getting further complicated by peephole passes that occur inbetween. What is the reason for this ordering?